reference/src/clientimpl/src/sa_key_provision.c

```diff
@@ -0,0 +1,364 @@
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "sa.h"
+#include "sa_rights.h"
+#include "log.h"
+#include <stdbool.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <string.h>
+#include <assert.h>
+
+sa_key *create_uninitialized_sa_key() {
+   sa_key *handle = (sa_key*)calloc(sizeof(sa_key), 1);
+   if(NULL == handle) {
+      ERROR("Failed to allcate memory");
+      return NULL;
+   }
+   *handle = INVALID_HANDLE;
+   return handle;
+}
+
+sa_status release_sa_key(sa_key *key) {
+   sa_status status = SA_STATUS_OK;
+
+   if(NULL != key) {
+      if(INVALID_HANDLE != *key) {
+         status = sa_key_release(*key);
+         if(SA_STATUS_OK != status) {
+            ERROR("Failed to release key, status:%d", status);
+         }
+      }
+      free(key);
+      key = NULL;
+   }
+   return status;
+}
+
+sa_status sa_key_provision_ta (
+        sa_ta_key_type ta_key_type,
+        const void* in,
+        size_t in_length,
+        void* parameters) {
+
+   sa_status status = SA_STATUS_OK;
+   WidevineOemProvisioning *wvProvision = NULL;
+   PlayReadyProvisioning   *prProvision = NULL;
+   NetflixProvisioning     *nflxProvision = NULL;
+
+   if(NULL == in ||
+      0 == in_length) {
+      ERROR("null input or input size is  0");
+      return SA_STATUS_NULL_PARAMETER;
+   }
+
+   switch(ta_key_type) {
+      case WIDEVINE_OEM_PROVISIONING:
+         {
+            DEBUG(" case WIDEVINE_OEM_PROVISIONING");
+
+            if(NULL == ((WidevineOemProvisioning*)in)->oemDevicePrivateKey      ||
+               0 == ((WidevineOemProvisioning*)in)->oemDevicePrivateKeyLength   ||
+               NULL == ((WidevineOemProvisioning*)in)->oemDeviceCertificate     ||
+               0 == ((WidevineOemProvisioning*)in)->oemDeviceCertificateLength) {
+               ERROR("null private key/certor or size is  0");
+               return SA_STATUS_NULL_PARAMETER;
+            }
+
+            wvProvision = calloc(sizeof(WidevineOemProvisioning), 1);
+            if(NULL == wvProvision) {
+               ERROR("OOM");
+               return SA_STATUS_INTERNAL_ERROR;
+            }
+
+            /*deep copy*/
+            wvProvision->oemDevicePrivateKeyLength =
+               ((WidevineOemProvisioning*)in)->oemDevicePrivateKeyLength;
+            wvProvision->oemDeviceCertificateLength =
+               ((WidevineOemProvisioning*)in)->oemDeviceCertificateLength;
+
+            wvProvision->oemDevicePrivateKey =
+               (void*)calloc(wvProvision->oemDevicePrivateKeyLength, 1);
+            if(NULL == wvProvision->oemDevicePrivateKey) {
+               ERROR("OOM");
+               return SA_STATUS_INTERNAL_ERROR;
+            }
+            memcpy(wvProvision->oemDevicePrivateKey,
+                   ((WidevineOemProvisioning*)in)->oemDevicePrivateKey,
+                   wvProvision->oemDevicePrivateKeyLength);
+
+            /*here just provide an examlpe to pass certification data,
```

Comment:
Let's inlcude the "TODO SoC Vendor" label everywhere the SoC vendor is expected to make changes.